### PR TITLE
feat(starter): implement a new starter bundle with OpenApi 3 support

### PR DESCRIPTION
**Pull request by huebner-j**
This change introduces a new version of the `sda-commons-server-starter` module. The change contains the support of OpenApi 3. By integrating this module only OpenApi 3 annotations are supported in REST endpoint definitions (follow the the links for further details [1][2])

Due to the support of OpenApi 3 there are also changes to the SdaPlatformBundle. The creation of a dropwizard application differs slightly when using OpenApi 3. The following example shows how an application is created (see [3]):

```java
...

@OpenAPIDefinition(
    info =
        @Info(
            title = "SDA Platform Example Application",
            description =
                "This application is an example to show a developer how to create a SDA Dropwizard application",
            version = "1.1.1",
            contact =
                @Contact(name = "John Doe", email = "info@example.com", url = "j.doe@example.com")))
public class SdaPlatformExampleApplication extends Application<MyCustomConfiguration> {

  public static void main(String[] args) throws Exception {
    new SdaPlatformExampleApplication().run(args);
  }

  @override
  public void initialize(Bootstrap<MyCustomConfiguration> bootstrap) {
    bootstrap.addBundle(
        SdaPlatformBundle.builder()
            .usingSdaPlatformConfiguration(MyCustomConfiguration.class)
            // Additional Swagger documentation properties may be set here until
            // the packages that should be scanned for Swagger documentation annotations are defined
	    .addOpenApiResourcePackageClass(this.getClass())
            // or use an existing existing OpenAPI definition.
            // provide the path to the existing openapi file (yaml or json) in the classpath
            .withExistingOpenAPIFromClasspathResource("/custom-openapi.yaml")
            .build());
  }

  ...
```

Hint: Swagger / OpenApi 2 can be used by importing the module `sda-commons-server-starter-swagger`

[1] - https://github.com/SDA-SE/sda-dropwizard-commons/tree/master/sda-commons-server-openapi
[2] - https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Annotations
[3] - https://github.com/SDA-SE/sda-dropwizard-commons/blob/master/sda-commons-server-starter-example/src/main/java/org/sdase/commons/server/starter/example/SdaPlatformExampleApplication.java

### CODE REVIEW

File: sda-commons-starter/build.gradle
Comment (dhenneke) [inline sda-commons-starter/build.gradle:16]: Is already imported by `sda-commons-server-testing`.

File: sda-commons-starter/build.gradle
Comment (dhenneke) [inline sda-commons-starter/build.gradle:9]: No `:` at the end.

File: sda-commons-starter/build.gradle
Comment (dhenneke) [inline sda-commons-starter/build.gradle:1]: Very minor: We could sort this.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (dhenneke) [inline sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:152]: Open for discussion: Since most applications derive their configuration from `SdaPlatformConfiguration` it would be nice to give them a shortcut like `public <T extends SdaPlatformConfiguration> AuthConfigProviderBuilder<T> usingSdaPlatformConfiguration(Class<T> configurationClass)`. I would also like to enable OPA by default in this builder. No OPA would require to use the `usingCustomConfig()` branch.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (dhenneke) [inline sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:233]: Open for discussion: Is there ever a need to make this configurable? I would also vote to remove it from the default `SdaPlatformConfiguration` class if it is not the default.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (dhenneke) [inline sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: Is this used often?

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (dhenneke) [inline sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:274]: Open for discussion: Do we prefer to do this in code or in config?

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (dhenneke) [inline sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:334]: I never seen someone using it and nobody complained about it in the last 19 months it is present, so we should remove the option.

Since the OpenAPI module is still "experimental" we might remove it from there too.

Comment (JoergSiebahn) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:334]: As the embed parameter is only documented if the response type has embeddable objects, this setting could lead to wrong documentation. In my opinion, it could be removed in both modules.

**Comment by Fox32** [comment]
I can't ignore the code duplications. Maybe we set it in gradle instead?

File: sda-commons-starter/build.gradle
Comment (huebner-j) [reply sda-commons-starter/build.gradle:16]: I've removed the dependency..

File: sda-commons-starter/build.gradle
Comment (huebner-j) [reply sda-commons-starter/build.gradle:9]: I have no clue how this came about.. Removed!

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: I will investigate..

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java]: If nobody is using this option, we should definitely remove it... Done :)

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (JoergSiebahn) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: I could not find the method name in the whole SDA GitHub ORG apart from the commons repo.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java]: >  In my opinion, it could be removed in both modules.

@JoergSiebahn isn't it a breaking change if we remove the method from the `sda-commons-server-starter` modul, even though nobody is actually using it?

Comment (JoergSiebahn) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java]: I think, we can safely remove it `disableEmbedParameter` from `OpenApiBundle` which is experimental and the new `SdaPlatformBundle` in `sda-commons-starter`.
You are right, we should NOT remove it in `sda-commons-server-starter` or `sda-commons-server-swagger`.
Maybe I should have clarified  _„both modules“_.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (Fox32) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:274]: I don't think that this is part of the deployment. This is part of the implementation. There shouldn't be any need to expose additional headers if the implementation doesn't require it. If you still want to override the default behavior of the implementation, use an ingress.

Comment (JoergSiebahn) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:274]: Some services in our org use it to allow _"Content-Range"_ for a list result instead of providing paging parameters in a response model. Do you know other cases? This example does not match our Rest Guide, I think.

Comment (Fox32) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:274]: `Content-Range` is not designed for paging as it requests specific bytes in a resource, but I won't judge this. My point is, that if the REST Api has such a feature, the REST API implementer has to allow the access to the header. We should let the consumer figure out that it doesn't work.

We had some implementations that expose the total count for paging as a header, but even there we have another recommendation in the REST guide.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:152]: I've created a PR for this change. Let's discuss the suggestion there. 
[https://github.com/SDA-SE/sda-dropwizard-commons/pull/654](https://github.com/SDA-SE/sda-dropwizard-commons/pull/654)

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: I couldn't find any use of this method either. I'll remove it from the builder..

Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: Team @SDA-SE/communications uses this method [here](https://github.com/SDA-SE/malware-scanner-facade/blob/e36436bd4531a2376e6ebcbce093142bb1897415/malware-scanner-service/src/main/java/com/sdase/malware/scanner/facade/MalwareScannerFacadeApp.java#L88). So the answer is, that it is not often used, it is used exactly once.

Comment (christopher-cudennec) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: ... at least in our organization. Can't speak for other customers.

Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: That's true. But I think, the decision was made by your use of the method, that we leave the method as it is.

Comment (JoergSiebahn) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:236]: Oops, I missed that match when looking into the [search results](https://github.com/search?q=org%3ASDA-SE+withExcludePatternsForRequiredConsumerToken&type=code). Thanks, that you checked it again.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (dhenneke) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:274]: I thought that this was also part of the configuration class but it isn't. So no need for further actions.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java]: Ok perfect, I've removed it from the `OpenApiBundle` & the `sda-commons-starter` builder..

File: sda-commons-starter/build.gradle
Comment (huebner-j) [reply sda-commons-starter/build.gradle:1]: What criteria do you like to sort the dependencies by? By name?

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [inline sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: Open for discussion: When I first used the `ConsumerToken` configuration in the starter bundle I was wondering why there is no `consumerToken` in the configuration until i understood that this consumer token configuration is just for making consumer tokens optional or required for incoming requests. I think the naming of the configuration is quite confusing since the name is `withConsumerTokenConfigProvider(SdaPlatformConfiguration::getConsumerToken)`. When I read `getConsumerToken` then I would expect the actual consumer token and not a configuration. Furthermore if you would like to define a consumer token for outgoing calls, you have to name it like 'consumerTokenName'. I would vote for a renaming here to make clear that this configuration is for incoming requests.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:233]: I've searched in our organization and found some occurrences:  
[1](https://github.com/SDA-SE/edithcare-participation-service/blob/fa61f6d544d833373d895721b3b803ad60ff4245/config.yml#L21)
[2](https://github.com/SDA-SE/customer-service/blob/fea7b0fb679a51d2835b0621940fc91ea31ab0fc/src/integTest/resources/test-config.yml#L16)
[3](https://github.com/SDA-SE/billing-external/blob/15fe3b544b7606f3dcfd68903789609a6f55f893/billing-external-service/config.yaml#L10)
And it seems some others are using it as well. [search in our organization](https://github.com/search?l=YAML&q=org%3ASDA-SE+consumerToken&type=Code)

I would also vote to remove it from the `SdaPlatformConfiguration`.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (christopher-cudennec) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: I've also heard from different people that the name is not intuitively clear.

Although the name is used according to other config properties. E.g. the mongo configuration is mostly added as "mongo".

What would be a better name? `consumerTokenSettings`? `consumerTokenServer`?

Comment (JoergSiebahn) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: The consumer token or consumer name seems like a workaround since I'm working for the SDA. At the moment it is only used for logging. I wonder if it is needed to set it technically as required in a service while it is a (temporary?) required pattern for the platform in general that may change when we introduce ISTIO.
It was meant to require identification of the calling service which should be part of the authorization we do in OPA. Actually it is nothing more than an indicator of who is calling as it can't be verified. I know moving that decision to OPA and not requiring it in the service may have a big impact on existing services but gives more flexibility for the future. I'd like to hear about @Fox32 opinion. I think he has quite a lot of insights about the requirements of the consumer token, the use in current services at SDA and at developing customers and about the future evolution of service authentication in context of a service mesh. Maybe @dhenneke can also point out some arguments about moving such a decision to OPA.

**Comment by huebner-j** [comment]
> I can't ignore the code duplications. Maybe we set it in gradle instead?

@Fox32 all files from `sda-commons-server-starter` are excluded from copy & paste detection by this property: [https://github.com/SDA-SE/sda-dropwizard-commons/blob/feat/new-starter-module/build.gradle#L190](https://github.com/SDA-SE/sda-dropwizard-commons/blob/feat/new-starter-module/build.gradle#L190). Is it fine for you?

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (dhenneke) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: My main concern with `withConsumerTokenConfigProvider(...)` is that we have two options to configure the consumer token: You can use `withoutConsumerTokenSupport()` / `withRequiredConsumerToken()` / `withOptionalConsumerToken()` OR the `withConsumerTokenConfigProvider(...)`. My current understanding of the consumer token (and I guess the majority selection) is that the service should decide whether it want to require a consumer token during build-time. We shouldn't give the users of the `SdaPlatformBundle` the choice between deciding it in code vs. deciding it in config, regardless of the decision whether it should be optional or required. Especially since there is no guidance on what the preferred way is. (And I agree that it is very easy to confuse the server-setting and the client-setting in a configuration).

About the need, our wiki cites:
> A consumer token is used to identify the consumer of a service. This is used to aggregate metrics by consumer, to filter logs, and for billing purposes. The consumer token should be provided with every request in form of the `Consumer-Token` request header. If no consumer token is provided, the access is denied with status `401 - UNAUTHORIZED`.

It's not really an actual authentication measure but more a technical/organizational guideline. In this regard, maybe the `401` was not the best choice, but that shouldn't be discussed in this PR.

Comment (christopher-cudennec) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: So your conclusion / recommendation is to remove `withConsumerTokenConfigProvider`?

Comment (dhenneke) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: Or remove the other three? Or we decide that this needs a deeper look and decide it in the future.

Comment (Fox32) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: As @dhenneke already said: Consumer-Token is only used for logging and metrics. So more or less a simple version of tracing. Maybe we could even make it completely optional and not require it at all. I even wonder how this works well with customers that don't use the consumer token in the infrastructure. That is probably also the reason for making it configurable at deployment time. Otherwise I would vote for making it an implementation time decision - so remove the option. However I'm not sure if the customers will like that.

Comment (JoergSiebahn) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java:229]: > Maybe we could even make it completely optional and not require it at all.

That would mean, we make _withOptionalConsumerToken()_ the default and remove all other options?

I would like that approach. We could add the other options if some requests for it.

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java]: Please take a look at [PR-662](https://github.com/SDA-SE/sda-dropwizard-commons/pull/662)

File: sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java
Comment (huebner-j) [reply sda-commons-starter/src/main/java/org/sdase/commons/starter/SdaPlatformBundle.java]: Please take a look at [PR-662](https://github.com/SDA-SE/sda-dropwizard-commons/pull/662)

**Review by JoergSiebahn — DISMISSED** [review]

File: sda-commons-starter-example/README.md
Comment (dhenneke) [inline sda-commons-starter-example/README.md:1]: As a next step, we should think about aligning this example nearer to our template-project and maybe also make an effort in improving the overall documentation, if needed.

File: sda-commons-starter/README.md
Comment (dhenneke) [inline sda-commons-starter/README.md:55]: We often store all rest-interfaces in a separate package that should also contain this definition. It is not always the best idea to register the application as a starting point. Sometimes the teams use generated client-code which also includes openapi-annotations. And we don't want these client interfaces be part of the OpenAPI file of the service.

But nothing we need to fix now.

File: sda-commons-starter/README.md
Comment (dhenneke) [inline sda-commons-starter/README.md:101]: This is a deprecated field. Might also be changed in a documentation-follow-up.

File: sda-commons-server-openapi/src/main/java/org/sdase/commons/server/openapi/OpenApiBundle.java
Comment (dhenneke) [inline sda-commons-server-openapi/src/main/java/org/sdase/commons/server/openapi/OpenApiBundle.java:228]: It would be nice to have this change in a separate commit with the correct `feat(server-openapi)` scope. Or can we leave it as-is @SDA-SE/sda-commons-maintainer?

**Review by dhenneke — COMMENTED** [review]
Thanks 🏆 

I would approve, but am happy to wait for the feedback from others.

PS: We should remember to add the experimental notice after the release has been created, similar to https://github.com/SDA-SE/sda-dropwizard-commons/releases/tag/2.3.0.

**Review by JoergSiebahn — APPROVED** [review]

File: sda-commons-server-openapi/src/main/java/org/sdase/commons/server/openapi/OpenApiBundle.java
Comment (christopher-cudennec) [reply sda-commons-server-openapi/src/main/java/org/sdase/commons/server/openapi/OpenApiBundle.java:228]: Would be nice to see it in an extra commit. Last request for change, promise! 😃

Comment (huebner-j) [reply sda-commons-server-openapi/src/main/java/org/sdase/commons/server/openapi/OpenApiBundle.java:228]: done :)

**Review by dhenneke — APPROVED** [review]

**Review by Fox32 — APPROVED** [review]
👍

**Pull request merged**
